File: 0031-next-permutation/0031-next-permutation.java
import java.util.Arrays;

public class Solution {
    public void nextPermutation(int[] nums) {
        int n = nums.length, piv = -1;

        // Step 1: Find pivot
        for (int i = n - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) {
                piv = i;
                break;
            }
        }

        // If no pivot is found, reverse the entire array
        if (piv == -1) {
            reverse(nums, 0, n - 1);
            return;
        }

        // Step 2: Find the smallest element greater than nums[piv]
        for (int i = n - 1; i > piv; i--) {
            if (nums[i] > nums[piv]) {
                swap(nums, i, piv);
                break;
            }
        }

        // Step 3: Reverse the suffix
        reverse(nums, piv + 1, n - 1);
    }

    private void swap(int[] nums, int a, int b) {
        int temp = nums[a];
        nums[a] = nums[b];
        nums[b] = temp;
    }

    private void reverse(int[] nums, int start, int end) {
        while (start < end) {
            swap(nums, start++, end--);
        }
    }
}

File: 0031-next-permutation/0031-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n=nums.size(),i,j,piv=-1;
        for(i=n-2;i>=0;i--)
        {
            if(nums[i]<nums[i+1])
            {
                piv=i;
                break;
            }
        }
        if(piv==-1)
        {
            reverse(nums.begin(),nums.end());
            return;
        }
        for(i=n-1;i>piv;i--)
        {
            if(nums[i]>nums[piv])
            {
                swap(nums[i],nums[piv]);
                break;
            }
        }
        i=piv+1,j=n-1;
        while(i<=j)
        {
            swap(nums[i],nums[j]);
            i+=1;
            j-=1;
        }
    }
};